pink/rondis/string/commands.cc

```diff
@@ -6,9 +6,56 @@
 #include <ndbapi/Ndb.hpp>
 
 #include "db_operations.h"
+#include "commands.h"
 #include "../common.h"
 #include "table_definitions.h"
 
+bool setup_transaction(
+    Ndb *ndb,
+    const pink::RedisCmdArgsType &argv,
+    std::string *response,
+    struct key_table *key_row,
+    const char *key_str,
+    Uint32 key_len,
+    const NdbDictionary::Dictionary **ret_dict,
+    const NdbDictionary::Table **ret_tab,
+    NdbTransaction **ret_trans)
+{
+    if (key_len > MAX_KEY_VALUE_LEN)
+    {
+        assign_generic_err_to_response(response, REDIS_KEY_TOO_LARGE);
+        return false;
+    }
+    const NdbDictionary::Dictionary *dict = ndb->getDictionary();
+    if (dict == nullptr)
+    {
+        assign_ndb_err_to_response(response, FAILED_GET_DICT, ndb->getNdbError());
+        return false;
+    }
+    const NdbDictionary::Table *tab = dict->getTable(KEY_TABLE_NAME);
+    if (tab == nullptr)
+    {
+        assign_ndb_err_to_response(response, FAILED_CREATE_TABLE_OBJECT, dict->getNdbError());
+        return false;
+    }
+    memcpy(&key_row->redis_key[2], key_str, key_len);
```

Comment:
Is this necessary? Since we should never change the key..

Reply:
We add two rows of NdbRecord, one is input that contains the searched for primary key, the second is the new values to write. We use the same row for both, so setting the primary key is part of the WHERE clause. Also it is used to write the values when performing an INSERT operation (the WRITE is either becoming an UPDATE or an INSERT.

Reply:
What I mean is whether the memory has to be copied or whether one could just reference the pointer

Reply:
You have defined an NdbRecord, this defines the offset from the pointer to the Primary key pointer row that you supply in writeTuple, so if you define the offset as 0 and send in key_str as the Primary key pointer I gather it could work, but for a multi-column PK you need to copy, I don't think this optimisation makes much sense, most keys are short and so not a problem to do a quick memcpy